### PR TITLE
[libfswatch] New port

### LINKED ISSUE #251: Vcpkg support (Windows package manager)
Hi, I was wondering if `vcpkg` integration could be fixed. Someone posted the issue more than a year ago in vcpkg's repo around a year ago: https://github.com/Microsoft/vcpkg/issues/5843.

According to the post, the problems preventing inclusion are:
> - libfswatch_config.h are not found in source.
> - This port only supports Linux.
> - Without build documentation.
> - Some macro such as PACKAGE_STRING PACKAGE_NAME PACKAGE_BUGREPORT PACKAGE PACKAGE_URL are not found.
> - It seems like that the port has code bug in source. eg: optarg and getopt not found.

There is currently a pull request on fixing CMAKE, but I haven't looked into it: https://github.com/emcrisostomo/fswatch/pull/250#issue-465250042

Related: https://github.com/emcrisostomo/fswatch/issues/176.

Thanks.

### PR DESCRIPTION
**Pull request by SamuelMarks**
**Describe the pull request**
> fswatch is a cross-platform file change monitor with multiple backends: Apple OS X File System Events, *BSD kqueue, Solaris/Illumos File Events Notification, Linux inotify, Microsoft Windows and a stat()-based backend. 

- #### What does your PR fix?  
  Fixes https://github.com/emcrisostomo/fswatch/issues/251

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  
  all

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
  WiP

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
I am still working on this PR

**If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/**
WiP

NOTE: I am pinning this to my fork rather than the official source as its maintainer hasn't been active since June (I opened the PR adding MSVC support in November). Tried emailing him. If he merged the PR then happy to switch the pin to his repo.

### CODE REVIEW

File: ports/libfswatch/vcpkg.json [hunk 1]
```diff
@@ -0,0 +1,21 @@
+{
+  "name": "libfswatch",
+  "version-string": "1.17.0",
```
Comment (JonLiu1993) [inline ports/libfswatch/vcpkg.json:3]:
```suggestion
  "version": "1.17.0",
```

File: ports/libfswatch/portfile.cmake [hunk 2]
```diff
@@ -0,0 +1,15 @@
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO SamuelMarks/fswatch
+    REF 6b273e1285cfb6cf6df3fbd4d593a07500184559
+    SHA512 4e13465fb2f51e68e3a37bc894415b3f1474144e12aad00e9c351ede04ad3ed5d58a2988c44d94a5589ec88373f6d80b7367c7b62c00e6f230c71a1f63fa064f
+    HEAD_REF multi-os-ci
+)
+
+vcpkg_cmake_configure(
+    SOURCE_PATH ${SOURCE_PATH}
```
Comment (JonLiu1993) [inline ports/libfswatch/portfile.cmake:10]:
```suggestion
    SOURCE_PATH "${SOURCE_PATH}"
```

File: ports/libfswatch/portfile.cmake [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+)
+vcpkg_cmake_install()
+file(INSTALL ${SOURCE_PATH}/COPYING
+     DESTINATION ${CURRENT_PACKAGES_DIR}/share/libfswatch
+     RENAME copyright)
```
Comment (JonLiu1993) [inline ports/libfswatch/portfile.cmake:15]:
```suggestion
file(INSTALL "${SOURCE_PATH}/COPYING"
     DESTINATION "${CURRENT_PACKAGES_DIR}/share/libfswatch"
     RENAME copyright)
```

**Comment by JonLiu1993** [comment]
@SamuelMarks ,Thanks for your pr, looks like ci test failed on all triplet
This is error log ,could you please take a look:
https://dev.azure.com/vcpkg/public/_build/results?buildId=65889&view=artifacts&pathAsName=false&type=publishedArtifacts
````
D:\buildtrees\libfswatch\src\aee8dab901-10131ccb62.clean\libfswatch\src\libfswatch\c++\path_utils.cpp(23): fatal error C1083: Cannot open include file: 'libfswatch_config.h': No such file or directory
````

**Comment by SamuelMarks** [comment]
@JonLiu1993 Yeah I've got a pretty extensive CMake refactor I'm working on, about to commit it up

(this PR is still a draft)

**Comment by SamuelMarks** [comment]
@JonLiu1993 Ready for review!

**Comment by SamuelMarks** [comment]
@JonLiu1993 Ready for review!

FYI: This should work on macOS, Windows, Linux; with the exception of UWP and Windows ARM (not sure why Windows ARM fails, maybe one of libfswatch's vcpkg-listed dependencies?).

File: ports/libfswatch/portfile.cmake [hunk 4]
```diff
@@ -0,0 +1,22 @@
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO            SamuelMarks/fswatch
+    REF             045f44c1c410f1f3c425e10d59c786a86401ccbd
+    SHA512          7333a33066b7dbf50304b405921b93229941455f70902e14162e3cc47e6822fa8cfa0ccce19bfebea73867f648fcdaeaab32ce48df9e01f4ef4149dac09b4dcb
+    HEAD_REF        multi-os-ci
```
Comment (JonLiu1993) [inline ports/libfswatch/portfile.cmake:6]:
```suggestion
    REPO SamuelMarks/fswatch
    REF 045f44c1c410f1f3c425e10d59c786a86401ccbd
    SHA512      7333a33066b7dbf50304b405921b93229941455f70902e14162e3cc47e6822fa8cfa0ccce19bfebea73867f648fcdaeaab32ce48df9e01f4ef4149dac09b4dcb
    HEAD_REF multi-os-ci
```

**Review by JonLiu1993 — DISMISSED** [review]

File: ports/libfswatch/vcpkg.json [hunk 5]
```diff
@@ -0,0 +1,27 @@
+{
+  "name": "libfswatch",
+  "version": "1.17.0",
+  "description": "A cross-platform file change monitor with multiple backends: Apple OS X File System Events, *BSD kqueue, Solaris/Illumos File Events Notification, Linux inotify, Microsoft Windows and a stat()-based backend.",
+  "supports": "!uwp & (arm64 & !windows)",
```
Comment (autoantwort) [inline ports/libfswatch/vcpkg.json:5]: It does not work on x64? And not on windows? I guess you probably mean `"!uwp & !(arm64 & windows)"` (I wounder why arm64 should not work).

**Review by strega-nil-ms — CHANGES_REQUESTED** [review]
@autoantwort is right, the supports clause is definitely wrong.

File: ports/libfswatch/vcpkg.json (hunk 5 above)
Comment (SamuelMarks) [reply ports/libfswatch/vcpkg.json:5]: @autoantwort / @strega-nil-ms Yeah I thought my clause was suspect (which is why I said it in plain-English). ARM64 Windows should work, the only thing I can think could be a stopper is if any of the dependencies listed in `vcpkg.json` don't support ARM64 Windows. UWP I think stops a lot of the internal Windows headers from working, so it's no surprise it fails there.

Comment (autoantwort) [reply ports/libfswatch/vcpkg.json:5]: > the only thing I can think could be a stopper is if any of the dependencies listed in vcpkg.json don't support ARM64 Windows

You don't have to care about this, simply remove the arm64 restriction. 

> UWP I think stops a lot of the internal Windows headers from working, so it's no surprise it fails there.

You can simply remove the supports expression and see what fails in reality

**Comment by strega-nil-ms** [comment]
Why is this called `libfswatch` and not `fswatch`?

**Comment by strega-nil-ms** [comment]
Looks like the correct `supports` expression is `"supports": "!(uwp | (windows & arm))"`

**Comment by SamuelMarks** [comment]
@strega-nil-ms Yeah the `#include <intrin.h>` was required for x86 support (and didn't effect x64 support). So should I just guard it to only `#include` if arch is x86?

**Comment by SamuelMarks** [comment]
@JonLiu1993 I think the CI needs to be kicked

**Comment by JonLiu1993** [comment]
> @JonLiu1993 I think the CI needs to be kicked

@SamuelMarks , I think you should consider @strega-nil-ms 's [suggestion](https://github.com/microsoft/vcpkg/pull/22527#issuecomment-1015983478)

**Comment by SamuelMarks** [comment]
@JonLiu1993 Nah I think I can fix ARM support, the CI showed these style errors:
```
:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um\winnt.h(8113): error C3861: 'ReadAcquire8': identifier not found
C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um\winnt.h(8124): error C3861: 'ReadNoFence8': identifier not found
C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um\winnt.h(8135): error C3861: 'ReadAcquire8': identifier not found
C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um\winnt.h(8146): error C3861: 'ReadNoFence8': identifier not found
C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um\winnt.h(8169): error C3861: 'WriteRelease8': identifier not found
C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um\winnt.h(8182): error C3861: 'WriteNoFence8': identifier not found
C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um\winnt.h(8195): error C3861: 'WriteRelease8': identifier not found
C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um\winnt.h(8208): error C3861: 'WriteNoFence8': identifier not found
C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um\winnt.h(8233): error C3861: 'ReadAcquire16': identifier not found
C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um\winnt.h(8244): error C3861: 'ReadNoFence16': identifier not found
C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um\winnt.h(8267): error C3861: 'WriteRelease16': identifier not found
C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um\winnt.h(8280): error C3861: 'WriteNoFence16': identifier not found
C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um\winnt.h(8305): error C3861: 'ReadAcquire': identifier not found
C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um\winnt.h(8316): error C3861: 'ReadNoFence': identifier not found
C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um\winnt.h(8339): error C3861: 'WriteRelease': identifier not found
C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um\winnt.h(8352): error C3861: 'WriteNoFence': identifier not found
C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um\winnt.h(8377): error C3861: 'ReadAcquire64': identifier not found
C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um\winnt.h(8388): error C3861: 'ReadNoFence64': identifier not found
C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um\winnt.h(8411): error C3861: 'WriteRelease64': identifier not found
C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um\winnt.h(8424): error C3861: 'WriteNoFence64': identifier not found
C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um\winnt.h(8561): error C3861: 'ReadAcquire64': identifier not found
C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um\winnt.h(8572): error C3861: 'ReadNoFence64': identifier not found
C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um\winnt.h(8595): error C3861: 'WriteRelease64': identifier not found
C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um\winnt.h(8608): error C3861: 'WriteNoFence64': identifier not found
C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um\winnt.h(20290): error C3861: '__iso_volatile_store8': identifier not found
C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um\winnt.h(22979): error C3861: '__getReg': identifier not found
C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um\winnt.h(22988): error C3861: '__getReg': identifier not found
```

Watch this space whilst I try and resolve them. Probably just a missing include or two, and existing includes which need guarding.

**Comment by SamuelMarks** [comment]
Actually on second thought it's not obvious why this is failing.

I was considering adding an explicit `#include <winnt.h>` to [`win_handle.cpp`](https://github.com/SamuelMarks/fswatch/blob/multi-os-ci/libfswatch/src/libfswatch/c++/windows/win_handle.cpp) but I really don't think that'll do anything.

Ideas anyone? - E.g, @strega-nil-ms / @strega-nil

**Comment by JonLiu1993** [comment]
@SamuelMarks , 
This is error log:
````
CMake Error in libfswatch/src/libfswatch/gettext/CMakeLists.txt:
  Target "gettext" INTERFACE_INCLUDE_DIRECTORIES property contains path:

    "D:/buildtrees/libfswatch/src/448f01fd84-a0f24a6984.clean/libfswatch/src/libfswatch/gettext/"

  which is prefixed in the source directory.


-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    CMAKE_INSTALL_BINDIR
    CMAKE_INSTALL_LIBDIR
    _VCPKG_ROOT_DIR


CMake Generate step failed.  Build files cannot be regenerated correctly.
````

**Comment by JonLiu1993** [comment]
Draft this pr because there is no response for a long time

**Comment by JonLiu1993** [comment]
This PR has been inactive for a long time, please reopen it if there is any progress.

**Pull request closed without merging**